File: 0340-longest-substring-with-at-most-k-distinct-characters/0340-longest-substring-with-at-most-k-distinct-characters.cpp
class Solution {
public:
    int getCnt(vector<int>& dict){
        int cnt=0;
        for(auto& i:dict) if(i) cnt++;
        return cnt;
    }
    int lengthOfLongestSubstringKDistinct(string s, int k) {
        vector<int> dict(256,0);
        int anc=0,res=0,n=s.length();
        for(int i=0;i<n;i++){
            dict[s[i]]++;
            while(getCnt(dict)>k && anc<=i){
                dict[s[anc]]--;
                anc++;
            }
            res=max(res,i-anc+1);
        }
        return res;
    }
};

File: 0340-longest-substring-with-at-most-k-distinct-characters/0340-longest-substring-with-at-most-k-distinct-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        # sliding window dict size <=2 that time reduce the anchor ptr 
        if k==0: 
            return 0 
        anc = 0 
        freq = defaultdict(int)
        ans = 0
        for i in range(len(s)): 
            freq[s[i]]+=1 
            while anc<i and len(freq)>k:
                freq[s[anc]]-=1 
                if freq[s[anc]]==0: 
                    del freq[s[anc]]
                anc+=1 
            ans = max(ans,i-anc+1) 
        return ans 
